o/snapstate: update default provider if missing required content

MiguelPires opened the pull request:
Updates a default provider when installing a snap if it's already installed but missing a required content attribute.

stolowski commented on overlord/snapstate/handlers.go:
```diff
@@ -286,6 +288,17 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+
+		if !ok {
+			// default provider is missing some content tags (likely outdated) so update it
+			return Update(st, snapName, nil, userID, flags)
```
I *think* we can hit store.ErrNoUpdateAvailable here if for whatever reason the content provider hasn't been updated in the store? Is this error going to surface and prevent entire install/update (would be change that breaks compatibility)?

MiguelPires replied:
Yes, that would prevent the installation. I incorrectly assumed that previously the installation would fail if the plug couldn't be auto-connected but I see that's not true (which makes sense). I can change it to log that snapd failed to update the prereq and proceed with the install

stolowski commented:
@MiguelPires the unit test failure might be related to the changes?

MiguelPires commented:
It is, the test is flaky because prerequisites are now stored in a map so the iteration order varies which makes the test check fail sometimes. I was waiting for https://github.com/snapcore/snapd/pull/10643 to fix it. I'll do it today

pedronis reviewed (dismissed):
we need code to deal with conflicts/ongoing operations on the snap that needs updating. On top of that we also need an explicit test about a situation where the snap to update is also part of the same update change, we need a test equivalent to:

snap refresh consumer-with-new-content-need outdated-producer

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -309,20 +324,121 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+
+// Checks for conflicting tasks. Returns true if the operation should be skipped. The error
+// can be a state.Retry if the operation should be retried later.
+func shouldSkipToAvoidConflict(task *state.Task, snapName string) (bool, error) {
```
afaict this could reuse linkSnapInFlight  tweaking it a bit (mainly to return the relevant change), instead of defining a full new helper

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -309,20 +324,121 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+	if !ok {
+		// default provider is missing some content tags (likely outdated) so update it
+		ts, err := Update(st, snapName, nil, userID, flags)
+		if err != nil {
```
as we do for the Install path, this needs to check whether the error is a conflict and return a Retry in that case

pedronis reviewed (commented):
the code seems to be going the right directions but some I think some reuse is possible and still missing some case handling, also there is a code conflict atm

MiguelPires commented:
Thanks for the feedback @pedronis, I addressed those issues and added a test for a conflicting task from a different change. 
(I see you already self-requested a review, thanks!)

pedronis commented on overlord/snapstate/conflict.go:
```diff
@@ -166,6 +166,13 @@ func CheckChangeConflictMany(st *state.State, instanceNames []string, ignoreChan
 		}
+
+		// task doesn't conflict if it can't run
+		switch task.Status() {
```
this is equivalent to task.Status().Ready()

the change seems ok to me but probably needs to be done as a separate PR with some associated tests

MiguelPires replied:
I wondered about doing it as a separate PR but opted not to since it was only triggered in my branch. I'll split it off into a separate PR with some better testing then

MiguelPires replied:
Now that I had another look at this, I think this change is wrong. The issue is just that two updates to the same snap can conflict which means that the second one is retried after 30s while the test expects the task runner to converge sooner than that. I had wondered if blocking the install of the "main" snap because the default provider is outdated was ok. It might be better to significantly shorten the retry period or simply not retry at all, to prevent the "soft" dependency of the content provider from blocking an install for 30 seconds

pedronis commented on overlord/snapstate/conflict.go:
```diff
@@ -166,6 +166,13 @@ func CheckChangeConflictMany(st *state.State, instanceNames []string, ignoreChan
+		// task doesn't conflict if it can't run
+		switch task.Status() {
+		case state.HoldStatus, state.ErrorStatus, state.DoneStatus, state.UndoneStatus:
+			continue
```
I would move this though just before the affectedSnaps code below

MiguelPires replied:
I haven't change this code in this branch but I opened another PR for this fix (which incorporates this feedback) with proper tests. Once that merges, I'll just rebase this on it and drop this commit

MiguelPires replied:
https://github.com/snapcore/snapd/pull/10767

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +336,115 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+		ts, err := Update(st, snapName, nil, userID, flags)
+		if err != nil {
+			if _, ok := err.(*ChangeConflictError); ok {
+				return nil, &state.Retry{After: prerequisitesRetryTimeout, Reason: fmt.Sprintf("update failed due to conflict: %s", err)}
```
I would leave out Reason as Install code does

pedronis replied:
this bit should be tested, probably worth checking how something similar is tested for Install

MiguelPires replied:
The latest commit adds more tests to cover the ways in which conflicts can manifest which covers this path as well

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -208,16 +212,25 @@ func linkSnapInFlight(st *state.State, snapName string) (bool, error) {
 				}
 				if snapsup.InstanceName() == snapName {
-					return true, nil
+					tasks = append(tasks, tc)
```
because of the conflict checks there shouldn't really be more than one non-ready link-snap task around for a given snap, so this could probably just return at the first hit instead of returning a []*state.Task

stolowski replied:
+1

MiguelPires replied:
Thanks, fixed this as well in one of the latest commits

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +336,115 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+			continue
+		}
+
+		contentAttr, ok := val.(string)
```
there's no test hitting this

MiguelPires replied:
Added tests to fully cover `hasAllContentAttrs`

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +336,115 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+		}
+	}
+
+	return true, nil
```
and this

pedronis reviewed (commented):
overall looks code, some detail comments

MiguelPires commented:
Thanks for the feedback. The latest commits expand the tests to fully cover hasAllContentAttrs and the conflicts cases in updatePrereqIfOutdated.  I didn't change the conflict bug fix because I've opened another PR for it so once that merges, this can also merge and I'll just drop the commit from this branch. I also slightly modified the TestUpdateContentProviderDownloadFailure test to not trigger two updates of the same snap just so the retry doesn't delay the task and cause the test to fail. IMO this change is fine since the test's purpose doesn't depend on it, it just depends on the consumer updating the producer

mardy commented on overlord/snapstate/handlers.go:
```diff
@@ -196,7 +198,7 @@ func defaultPrereqSnapsChannel() string {
 }
 
-func linkSnapInFlight(st *state.State, snapName string) (bool, error) {
+func inFlightLinkSnap(st *state.State, snapName string) (*state.Task, error) {
```
Maybe rename to `[find]LinkTaskForSnap()`?

I'm not sure what "in flight" means in this context, and whether it's helpful to have it in the function name.

MiguelPires replied:
I think it's meant to imply that it only looks for a task that can still run, i.e., a non-Ready() task. But I agree, it's not very clear

mardy commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +334,112 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+		}
+
+		t.Logf("installed prerequisite '%s' is missing content. Failed to update: %s", snapName, err)
+		return nil, nil
```
If it's intentional that we don't return an error, we should have a comment explaining why.

MiguelPires commented:
Thanks everyone for the feedback. I reverted the previous code that dealt with the issue of the DownloadFailure the retry and I did a series of master merges to get the code I needed plus the code to unblock the github actions/unit testing. As for the actual new code, I added the check that skips the update if there's a duplicate one conflicting. I added a test as well to cover this scenario plus a bit of refactoring/clarifying other tests

stolowski commented on overlord/snapstate/handlers.go:
```diff
@@ -208,16 +210,25 @@ func linkSnapInFlight(st *state.State, snapName string) (bool, error) {
+	return nil, nil
+}
+
+func isLinkSnapInFlight(st *state.State, snapName string) (bool, error) {
```
This helper does very little / isn't much different than just using findLinkSnapTask() directly, and appears to be used only once, so maybe just drop it and use findLinkSnapTask?

stolowski commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +334,120 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+func updatePrereqIfOutdated(t *state.Task, snapName string, contentAttrs []string, userID int, flags Flags) (*state.TaskSet, error) {
+	st := t.State()
+
+	if len(contentAttrs) == 0 {
```
Nitpick^2, this could go before st := ...

stolowski commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +334,120 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+
+		// don't propagate error to avoid failing the main install since the
+		// content provider is (for now) a soft dependency
+		t.Logf("installed prerequisite '%s' is missing content. Failed to update: %s", snapName, err)
```
I wonder if we should maybe reorder the messages here to say that update of prereq failed with err, and therefore it's content is missing?

MiguelPires replied:
Hm yes, I think that would be better since it'd make the situation clearer for the user. I was trying to imply that there was a prerequisite with missing content and that we tried but failed to update it but that that's not very relevant for the user and might just confuse them

stolowski commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +323,120 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+	if ok, err := hasAllContentAttrs(st, snapName, contentAttrs); err != nil {
+		return nil, err
+	} else if ok {
+		return nil, nil
```
I think we are not hitting 'ok' case in the tests?

MiguelPires replied:
No, because if hard to deterministically hit this case, but there is a `TestHasAllContentAttributes` in `snapstate_install_test.go` that should cover all of it. The issue is that the `ok` case will only be hit when the content is missing before the update tasks are scheduled but is then updated before the handler runs

stolowski replied:
Wouldn't this case be hit when the provider snap has all the required content attributes and doesn't need updating, so after Install("consumer") we should see exactly one download happening for the consumer snap and not for the provider?

MiguelPires replied:
The prerequisites->content map in the prereq task is only populated with content attributes that aren't already provided (I added some comments clarifying that in defaultProviderContentAttrs in an earlier commit). To trigger this case, the content provider can't already have this content at first so the prerequisites task gets a map with the content attributes and default providers to install, but before the prerequisites task of the consumer gets to run the producer is updated. This can happen because an update of the producer would not be prevent since the consumer's prerequisite task hasn't run and the producer install/update tasks haven't been scheduled. I couldn't think of a test that triggers this deterministically

stolowski commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +323,120 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+	ts, err := Update(st, snapName, nil, userID, flags)
+	if err != nil {
+		if conflErr, ok := err.(*ChangeConflictError); ok {
+			// there's already an update for the same snap in this change,
+			// just skip this one
+			if conflErr.ChangeID == t.Change().ID() {
```
AFAIU (but I may be missing something) we generally avoid conflict with shouldSkipToAvoidConflict() above, so I'm not quite clear what are the exact circumstances we can hit it here (even considering that Update unlocks the state around store call) - but I can see we definitely hit this case in the unit tests. I think both this and the call to shouldSkipToAvoidConflict above should have a comment explaining the conflicts, and why can we hit it again here.

MiguelPires replied:
It's been a while since the previous check was discussed so I may be forgetting something but, looking at it, it seems like what `shouldSkipToAvoidConflict` does is a subset of the conflict checking in Update so the first could be removed. Maybe @pedronis can correct me if I'm wrong

pedronis replied:
Update talks to the store first, so I think this is fine as is

mardy reviewed (approved):
LGTM, although it's hard to review it properly due to the size. :-)

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +323,122 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+	}
+
+	// TODO: this check is likely subset of the conflict checking in Update.
+	// Consider removing it
```
we should drop the TODO, I looked at the code again but Update basically checks for conflicts only after having consulted the store, that's a slow process, so this is just an optimisation in a sense, but I think we do want it

MiguelPires replied:
Make sense, I can replace the TODO with a small comment indicating it's an optimization (just to avoid confusion in the future)

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -310,20 +323,122 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
+		if conflErr, ok := err.(*ChangeConflictError); ok {
+			// there's already an update for the same snap in this change,
+			// just skip this one
+			if conflErr.ChangeID == t.Change().ID() {
```
I'm not sure it's an easy scenario to reach but I think we want to add the same check to the Install in installOneBaseOrRequired when it checks the ChangeConflictError. Can be done in a small follow-up

pedronis reviewed (approved):
thanks, I haven't looked super closely to the tests, maybe @stolowski can do that.

I think we can drop the TODO about the extra conflict check. I commented on this.

We probably want a follow-up where:

* we add the same change conflict check to the Install path
*  I think because of remodel (this code is older than that), both the Install and the Update
   should actually become InstallWithDeviceContext and UpdateWithDeviceContext using the context from the task via DeviceCtx

stolowski commented on overlord/snapstate/handlers.go:
```diff
@@ -263,14 +265,26 @@ func (m *SnapManager) doPrerequisites(t *state.Task, _ *tomb.Tomb) error {
-	if err := m.installPrereqs(t, base, snapsup.Prereq, snapsup.UserID, perfTimings); err != nil {
+	// if a previous version of snapd persisted Prereq only, fill the contentAttrs.
+	// There will be not content attrs, so it will not update an outdated default provider
+	if len(snapsup.PrereqContentAttrs) == 0 && len(snapsup.Prereq) != 0 {
```
I think we are not hitting this case in tests.

stolowski reviewed (approved):
Looks good, but I think we're missing a test case for coming from old snapd that didn't populate PrereqContentAttrs? Can be a followup.
